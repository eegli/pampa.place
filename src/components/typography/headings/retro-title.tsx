type RetroTitleProps = {
  primary: string;
  secondary?: string;
  margin?: string;
};

export const FancyRetroTitle = ({
  primary,
  secondary,
  margin,
}: RetroTitleProps) => (
  <div
    style={{
      display: 'flex',
      margin: margin || '2rem auto',
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column',
      fontSize: 'clamp(10px, 2vw, 15px)',
    }}
  >
    <h1
      style={{
        position: 'relative',
        fontFamily: 'Exo',
        fontSize: '7em',
        margin: 0,
        transform: 'skew(-15deg)',
        letterSpacing: '0.03em',
      }}
    >
      <span
        style={{
          textShadow: '0 0 0.1em #8ba2d0, 0 0 0.2em black, 0 0 5em #165ff3',
        }}
      >
        {primary}
      </span>
      <span
        style={{
          position: 'absolute',
          left: 0,
          top: 0,
          backgroundImage:
            'linear-gradient(#032d50 25%, #00a1ef 35%, white 50%,   #20125f 50%,#8313e7 55%, #ff61af 75%  )',
          WebkitTextStroke: '0.01em #94a0b9',
          WebkitBackgroundClip: 'text',
          color: 'transparent',
        }}
      >
        {primary}
      </span>
    </h1>
    <h2
      style={{
        fontFamily: 'Mr Dafoe',
        fontWeight: 400,
        margin: 0,
        fontSize: '4.5em',
        marginTop: '-1.1em',
        color: 'white',
        textShadow: '0 0 0.05em #fff, 0 0 0.2em #fe05e1, 0 0 0.3em #fe05e1',
        transform: 'rotate(-7deg)',
      }}
    >
      {secondary}
    </h2>
  </div>
);
