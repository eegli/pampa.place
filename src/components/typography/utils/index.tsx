import Box from '@mui/material/Box';

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
