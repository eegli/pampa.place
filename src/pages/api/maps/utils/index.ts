export {apiData as data} from '@/config/maps';
import {NextApiHandler} from 'next';

type ErrorResponse = {
  error: string;
};

type DataResponse<T> = T extends unknown[]
  ? {
      data: T;
    }
  : {
      data: T[];
    };

export type ApiJSONHandler<T = unknown> = NextApiHandler<
  DataResponse<T> | ErrorResponse
>;
export type ApiTextHandler = NextApiHandler<string> | ErrorResponse;

export function nthQuery<T>(arg: T | T[], index = 0): T | undefined {
  return Array.isArray(arg) ? arg[index] : arg;
}
