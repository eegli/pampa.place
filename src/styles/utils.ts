import {keyframes} from '@mui/material/styles';

type KeyedColorFadeProps = [number, string][];

export const keyedColorFade =
  (prop: string) =>
  (...steps: KeyedColorFadeProps) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };
