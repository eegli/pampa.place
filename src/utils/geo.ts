import {LatLngLiteral} from '@/config/types';
import Tdistance from '@turf/distance';
import {BBox, MultiPolygon, point, Polygon, Units} from '@turf/helpers';
import TpointsWithinPolygon from '@turf/points-within-polygon';
import {randomPoint as TrandomPoint} from '@turf/random';

export function randomPointInMap(
  bbox: BBox,
  polygon: Polygon | MultiPolygon
): LatLngLiteral {
  do {
    const random = TrandomPoint(20, {bbox});
    const ptsWithin = TpointsWithinPolygon(random, polygon);
    if (ptsWithin.features.length) {
      const pt = ptsWithin.features[0].geometry.coordinates;
      return {lng: pt[0], lat: pt[1]};
    }
  } while (true);
}

// Area in square kilometers, distance in meters
// https://www.desmos.com/calculator/xlzbhq4xm0
export function calcScore(area: number, dist: number): number {
  if (dist < 0) return 0;
  const c = Math.log(area * dist + 1) * Math.sqrt(area * dist);
  const score = 5000 * Math.E ** -(dist / c);
  return Math.round(score);
}
// Returns distance between two points in kilometers

export function calcDist(
  p1: LatLngLiteral,
  p2: LatLngLiteral,
  units: Units = 'meters'
): number {
  const from = point([p1.lng, p1.lat]);
  const to = point([p2.lng, p2.lat]);
  return Tdistance(from, to, {units});
}
