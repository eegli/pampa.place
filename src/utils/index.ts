// Area in square kilometers, distance in meters
import { LatLngLiteral } from '@config';
import * as turf from '@turf/turf';

// https://www.desmos.com/calculator/xlzbhq4xm0
export function calculateScore(area: number, dist: number) {
  if (dist < 0) return 0;
  const c =
    Math.log(area + 1) * Math.sqrt(dist * area) + Math.log(dist ** 2 + 1);
  const score = 5000 * Math.E ** -(dist / c);

  return Math.round(score);
}
// Returns distance between two points in kilometers

export function calculateDistance(
  p1: LatLngLiteral,
  p2: LatLngLiteral,
  units: turf.Units = 'kilometers'
) {
  const from = turf.point([p1.lng, p1.lat]);
  const to = turf.point([p2.lng, p2.lat]);
  return turf.distance(from, to, { units });
}
// Formats distance
export function formatDist(km: number, toFixed = 1) {
  if (km < 0) return '-';
  if (km > 1000) {
    return (km / 1000).toFixed(toFixed) + ' m';
  }
  return km.toFixed(toFixed) + ' km';
}

export function max(num: number, limit: number) {
  return num >= limit ? limit : num;
}

export function id(): string {
  const array = new Uint32Array(8);
  window.crypto.getRandomValues(array);
  let str = '';
  for (let i = 0; i < array.length; i++) {
    str += (i < 2 || i > 5 ? '' : '-') + array[i].toString(16).slice(-4);
  }
  return str;
}
