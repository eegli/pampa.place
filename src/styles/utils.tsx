import Box from '@mui/material/Box';
import {keyframes} from '@mui/material/styles';

type KeyedColorFadeProps = [number, string][];

export const keyedColorFade =
  (prop: string) =>
  (...steps: KeyedColorFadeProps) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
