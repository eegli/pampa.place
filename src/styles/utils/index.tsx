import createCache from '@emotion/cache';
import Box from '@mui/material/Box';
import {keyframes} from '@mui/material/styles';

const isBrowser = typeof document !== 'undefined';

// On the client side, Create a meta tag at the top of the <head> and set it as insertionPoint.
// This assures that MUI styles are loaded first.
// It allows developers to easily override MUI styles with other styling solutions, like CSS modules.
export function createEmotionCache() {
  let insertionPoint;

  if (isBrowser) {
    const emotionInsertionPoint = document.querySelector<HTMLMetaElement>(
      'meta[name="emotion-insertion-point"]'
    );
    insertionPoint = emotionInsertionPoint ?? undefined;
  }

  return createCache({key: 'mui-style', insertionPoint});
}

export const keyedColorFade =
  (prop: string) =>
  (...steps: [number, string][]) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };

export const colorize = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
