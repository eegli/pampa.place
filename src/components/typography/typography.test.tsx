import {render} from '@/tests/utils';
import React from 'react';
import {FancyRetroTitle} from './retro-title';

describe('Typography', () => {
  it('renders retro title', () => {
    const {container} = render(
      <FancyRetroTitle primary="Retro" secondary="Title" />
    );
    expect(container.firstChild?.lastChild).toMatchSnapshot();
  });
});
