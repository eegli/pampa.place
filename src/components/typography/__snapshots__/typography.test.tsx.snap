// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography renders retro title 1`] = `
<div
  style="display: flex; margin: 2rem auto; align-items: center; justify-content: center; flex-direction: column;"
>
  <h1
    style="position: relative; font-family: Exo; font-size: 7em; margin: 0px; transform: skew(-15deg); letter-spacing: 0.03em;"
  >
    <span
      style="text-shadow: 0 0 0.1em #8ba2d0, 0 0 0.2em black, 0 0 5em #165ff3;"
    >
      Retro
    </span>
    <span
      style="position: absolute; left: 0px; top: 0px; color: transparent;"
    >
      Retro
    </span>
  </h1>
  <h2
    style="font-family: Mr Dafoe; font-weight: 400; margin: -1.1em 0px 0px 0px; font-size: 4.5em; color: white; text-shadow: 0 0 0.05em #fff, 0 0 0.2em #fe05e1, 0 0 0.3em #fe05e1; transform: rotate(-7deg);"
  >
    Title
  </h2>
</div>
`;
